opm/simulators/wells/BlackoilWellModelConstraints.hpp

```diff
@@ -0,0 +1,88 @@
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_BLACKOILWELLMODEL_CONSTRAINTS_HEADER_INCLUDED
+#define OPM_BLACKOILWELLMODEL_CONSTRAINTS_HEADER_INCLUDED
+
+#include <opm/input/eclipse/Schedule/Group/Group.hpp>
+
+namespace Opm {
+
+class BlackoilWellModelGeneric;
+class DeferredLogger;
+class GroupState;
+class SummaryState;
+class WellState;
+
+/// Class for handling constraints for the blackoil well model.
+class BlackoilWellModelConstraints
+{
+public:
+    BlackoilWellModelConstraints(const BlackoilWellModelGeneric& well)
+        : well_(well)
+    {}
+
+    /// Return true if any well has a THP constraint.
+    bool hasTHPConstraints() const;
+
+    std::pair<Group::InjectionCMode, double>
+    checkGroupInjectionConstraints(const Group& group,
+                                   const int reportStepIdx,
+                                   const Phase& phase) const;
+
+    std::pair<Group::ProductionCMode, double>
+    checkGroupProductionConstraints(const Group& group,
+                                    const int reportStepIdx,
+                                    DeferredLogger& deferred_logger) const;
+
+    bool checkGroupConstraints(const Group& group,
+                               const int reportStepIdx,
+                               DeferredLogger& deferred_logger) const;
+
+
+    void actionOnBrokenConstraints(const Group& group,
+                                   const Group::ExceedAction& exceed_action,
+                                   const Group::ProductionCMode& newControl,
+                                   GroupState& group_state,
+                                   DeferredLogger& deferred_logger) const;
+
+    void actionOnBrokenConstraints(const Group& group,
+                                   const Group::InjectionCMode& newControl,
+                                   const Phase& controlPhase,
+                                   GroupState& group_state,
+                                   DeferredLogger& deferred_logger) const;
+
+    bool updateGroupIndividualControl(const Group& group,
+                                      const int reportStepIdx,
+                                      std::map<std::pair<std::string,Opm::Phase>,std::string>& switched_inj,
+                                      std::map<std::string, std::string>& switched_prod,
+                                      GroupState& group_state,
+                                      WellState& well_state,
+                                      DeferredLogger& deferred_logger) const;
+
+private:
+    const BlackoilWellModelGeneric& well_;
```

Comment:
Should probably rename this object to something like `wellModel_`.

Reply:
done.